chanrestore.go

```diff
@@ -48,19 +48,40 @@ type chanDBRestorer struct {
 func (c *chanDBRestorer) openChannelShell(backup chanbackup.Single) (
 	*channeldb.ChannelShell, error) {
 
-	// First, we'll also need to obtain the private key for the shachain
-	// root from the encoded public key.
-	//
-	// TODO(roasbeef): now adds req for hardware signers to impl
-	// shachain...
-	privKey, err := c.secretKeys.DerivePrivKey(backup.ShaChainRootDesc)
-	if err != nil {
-		return nil, fmt.Errorf("unable to derive shachain root key: %v", err)
-	}
-	revRoot, err := chainhash.NewHash(privKey.Serialize())
-	if err != nil {
-		return nil, err
+	// The shachain root that seeds the RevocationProducer for this channel.
+	// It currently has two possible formats.
+	var revRoot *chainhash.Hash
+	var err error
+
+	// If the PubKey field is non-nil, then this shachain root is using the
+	// legacy non-ECDH scheme.
+	if backup.ShaChainRootDesc.PubKey != nil {
+		// Obtain the private key for the shachain root from the encoded public
+		// key.
+		privKey, err := c.secretKeys.DerivePrivKey(backup.ShaChainRootDesc)
+		if err != nil {
+			return nil, err
+		}
+
+		revRoot, err = chainhash.NewHash(privKey.Serialize())
+		if err != nil {
+			return nil, err
+		}
+	} else {
+		// This is the scheme in which the shachain root is derived via an
+		// ECDH operation on the private key of ShaChainRootDesc and the remote
+		// node's public key.
+		ecdh, err := c.secretKeys.ECDH(backup.ShaChainRootDesc, backup.RemoteNodePub)
+		if err != nil {
+			return nil, fmt.Errorf("unable to derive shachain root: %v", err)
+		}
+
+		revRoot, err = chainhash.NewHash(ecdh[:])
```

Comment:
We don't actually need to hash this _again_, since the x-coordinate which is the output of the ECDH is hashed before returning the output.

Reply:
Right, it's already a hash

Reply:
This will give us the wrong seed as implemented.

Reply:
This doesn't hash it as far as I can tell. This just converts it to `Hash`